ShopTask/Controllers/OrderController.cs

```
@@ -55,16 +54,19 @@ public async Task<JsonResult> GetProductsAsync()
         }
 
         [HttpPost]
-        public async Task<JsonResult> SaveOrder(Order order)
+        public async Task<JsonResult> SaveOrder(OrderModel orderModel)
         {
-            order.OrderItems = (OrderItem[])Session["Cart"];
-            if (!order.OrderItems.Any())
+            var order = Mapper.Map<OrderModel, Order>(orderModel);
+            order.OrderItems = Mapper.Map<OrderItemModel[], OrderItem[]>((OrderItemModel[])Session["Cart"]);
+            if (!(order.OrderItems?.Any() ?? new bool?(false)).Value)
             {
                 return Json(false, JsonRequestBehavior.AllowGet);
             }
+
+            var products = await _productsRepository.GetAllAsync(product => product.Category);
             foreach (var item in order.OrderItems)
             {
-                item.Price = (await _productsRepository.GetByIdAsync(item.ProductId)).Price;
+                item.Price = products.First(p => p.Id == item.ProductId).Price;
```

Принцип fail fast - используй products.Single - т к в норме всегда должен быть только один продукт с таким айди, если их оказалось несколько, лучше узнать об этом как можно раньше.